quill/ecs/src/entity_builder.rs

```diff
@@ -0,0 +1,123 @@
+use std::{
+    mem::{size_of, MaybeUninit},
+    ptr::{self, NonNull},
+};
+
+use crate::{component::ComponentMeta, Component, ComponentTypeId, Ecs, EntityId};
+
+/// A utility to build an entity's components.
+///
+/// An `EntityBuilder` can be reused to avoid repeated allocations.
+#[derive(Default)]
+pub struct EntityBuilder {
+    /// Packed vector containing component data.
+    components: Vec<MaybeUninit<u8>>,
+    entries: Vec<Entry>,
+}
+
+impl EntityBuilder {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Inserts a new component for the entity.
+    ///
+    /// If the entity builder already contains the component,
+    /// then the previous value is overriden.
+    pub fn add<T: Component>(&mut self, component: T) -> &mut Self {
```

Comment:
`Component` does not require `Copy` which means `T` could contain ptrs which will be dropped at the end of add?

Reply:
We wrap `component` in a `MaybeUninit` to prevent it from being dropped.

Reply:
Components are bound on Sync.

Reply:
Only `Send`, not `Sync`, right now. `Ecs` doesn't implement `Sync` as a result.

Reply:
I typed it out before checking and saw `Send` and sent the comment. But the idea behind it is still correct.

Reply:
He's referring to structs like `Vec` which hold pointers - if they're dropped twice, it's UB because of a double free.

Reply:
We still have to call `drop` on clear `EntityBuilder::clear` this was my roundabout way of getting there =)